modules/basics/src/main/java/com/opengamma/basics/date/ImmutableHolidayCalendar.java

```diff
@@ -141,11 +156,33 @@ private ImmutableHolidayCalendar(String name, SortedSet<LocalDate> holidays, Set
     this.holidays = ImmutableSortedSet.copyOfSorted(holidays);
     this.weekendDays = Sets.immutableEnumSet(weekendDays);
     if (holidays.isEmpty()) {
-      range = LocalDateRange.ALL;
+      this.range = LocalDateRange.ALL;
+      this.startYear = 0;
+      this.lookup = new int[0];
     } else {
       this.range = LocalDateRange.ofClosed(
           holidays.first().with(TemporalAdjusters.firstDayOfYear()),
           holidays.last().with(TemporalAdjusters.lastDayOfYear()));
+      this.startYear = range.getStart().getYear();
+      int[] array = new int[(range.getEndExclusive().getYear() - startYear) * 12];
+      for (LocalDate date : holidays) {
+        int index = (date.getYear() - startYear) * 12 + date.getMonthValue() - 1;
+        array[index] |= (1 << (date.getDayOfMonth() - 1));
+      }
+      LocalDate date = LocalDate.of(startYear, 1, 1);
+      for (int i = 0; i < array.length; i++) {
+        int monthLen = date.lengthOfMonth();
+        for (int j = 0; j < monthLen; j++) {
+          if (weekendDays.contains(date.getDayOfWeek())) {
```

Comment:
This code cannot assume a Sat/Sun weekend. It could be Sun, Mon and Thu! Doing the calcs to derive the 7 day pattern based on an arbitrary weekend input would be costly. Since these are cached, this shouldn't be a slowdown point.